pkg/network/node/iptables.go

```diff
@@ -227,7 +227,7 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
 
 func (n *NodeIPTables) ensureEgressIPRules(egressIP, mark string) error {
 	for _, cidr := range n.clusterNetworkCIDR {
-		_, err := n.ipt.EnsureRule(iptables.Prepend, iptables.TableNAT, iptables.Chain("OPENSHIFT-MASQUERADE"), "-s", cidr, "-m", "mark", "--mark", mark, "-j", "SNAT", "--to-source", egressIP)
+		_, err := n.ipt.EnsureRule(iptables.Prepend, iptables.TableNAT, iptables.Chain("OPENSHIFT-MASQUERADE"), "-s", cidr, "!", "-d", cidr, "-m", "mark", "--mark", mark, "-j", "SNAT", "--to-source", egressIP)
```

Comment:
This doesn't deal correctly with multiple cluster CIDRs.
eg, if your cluster network is using `10.128.0.0/14` and `10.192.0.0/14`, then this change will make it so that a packet from `10.128.0.0/14` to `10.128.0.0/14` wouldn't get NATted, and a packet from `10.192.0.0/14` to `10.192.0.0/14` wouldn't get NATted, but a packet from `10.128.0.0/14` to `10.192.0.0/14` or vice-versa would

Reply:
This also doesn't seem to address the bug; note that the destination in your example is `10.0.116.7`, which is NOT inside `10.128.0.0/14`. Thus, the packet would still get SNATted anyway. (AWS/GCP/Azure clusters by default use `10.[128-131].x.y` for pods and `10.0.x.y` for nodes/LBs.)

Reply:
>>> This doesn't deal correctly with multiple cluster CIDRs.
True. I completely missed that and will get it fixed.

>>>This also doesn't seem to address the bug; note that the destination in your example is 10.0.116.7, which is NOT inside 10.128.0.0/14. Thus, the packet would still get SNATted anyway. (AWS/GCP/Azure clusters by default use 10.[128-131].x.y for pods and 10.0.x.y for nodes/LBs.)
I agree with 10.0.116.7, not being inside 10.128.0.0/14 but in PREROUTING this gets changed (D-NAT) to the IP of a router pod, one being 10.130.4.33:
Chain KUBE-SERVICES (2 references)
target     prot opt source               destination
KUBE-FW-HEVFQXAKPPGAL4BV  tcp  --  0.0.0.0/0            10.0.116.7           /* openshift-ingress/router-default:http loadbalancer IP */ tcp dpt:80
KUBE-FW-MBAZS3WDHL45BPIZ  tcp  --  0.0.0.0/0            10.0.116.7           /* openshift-ingress/router-default:https loadbalancer IP */ tcp dpt:443
[...]
Chain KUBE-SEP-GMF6HD62GMXKGYYH (1 references)
target     prot opt source               destination
KUBE-MARK-MASQ  all  --  10.130.4.33          0.0.0.0/0            /* openshift-ingress/router-default:http */
DNAT       tcp  --  0.0.0.0/0            0.0.0.0/0            /* openshift-ingress/router-default:http */ tcp to:10.130.4.33:80
Am I missing something?